### PR TITLE
fix origin of erc20 xtokens transfer

### PR DESCRIPTION
**Pull request by zqhxuyuan**
relates to: https://github.com/AcalaNetwork/Acala/issues/2167#issuecomment-1144845273

Notice: current when transfer back ERC20 back to our chain, we should make sure there're enough native token for sibling parachain account to charge storage fee.

### CODE REVIEW

File: modules/currencies/src/lib.rs [hunk 1]
```diff
@@ -333,8 +333,16 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
...
+				let erc20_holding_account = T::AddressMapping::get_account_id(&T::Erc20HoldingAccount::get());
+				let origin_address = if erc20_holding_account == from.clone() {
+					T::AddressMapping::get_or_create_evm_address(to)
+				} else if erc20_holding_account == to.clone() {
+					T::AddressMapping::get_or_create_evm_address(from)
+				} else {
+					let origin = T::EVMBridge::get_origin().ok_or(Error::<T>::RealOriginNotFound)?;
+					T::AddressMapping::get_or_create_evm_address(&origin)
+				};
 				let address = T::AddressMapping::get_or_create_evm_address(to);
```
Comment (zqhxuyuan) [inline modules/currencies/src/lib.rs:346]: I'm not sure this is a good solution to fix `RealOriginNotFound` issue

File: modules/currencies/src/lib.rs [hunk 2]
```diff
@@ -333,8 +333,16 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
...
+				// withdraw to or deposit from erc20 holding account
+				let erc20_holding_account = T::AddressMapping::get_account_id(&T::Erc20HoldingAccount::get());
+				let origin_address = if erc20_holding_account == from.clone() {
+					T::AddressMapping::get_or_create_evm_address(to)
+				} else if erc20_holding_account == to.clone() {
+					T::AddressMapping::get_or_create_evm_address(from)
+				} else {
+					let origin = T::EVMBridge::get_origin().ok_or(Error::<T>::RealOriginNotFound)?;
+					T::AddressMapping::get_or_create_evm_address(&origin)
+				};
```
Comment (zqhxuyuan) [inline modules/currencies/src/lib.rs:345]: I'm not sure this is a good solution to fix `RealOriginNotFound` issue

File: modules/evm/src/lib.rs [hunk 3]
```diff
@@ -1774,6 +1776,10 @@ impl<T: Config> EVMTrait<T::AccountId> for Pallet<T> {
 	fn set_origin(origin: T::AccountId) {
 		ExtrinsicOrigin::<T>::set(Some(origin));
 	}
+
+	fn clear_origin() {
+		ExtrinsicOrigin::<T>::kill();
+	}
```
Comment (zjb0807) [inline modules/evm/src/lib.rs:1782]: It was removed here.
https://github.com/AcalaNetwork/Acala/blob/3c25a51339bfda1391dd9c541a8f9e70f1187812/modules/evm/src/lib.rs#L1914

File: modules/currencies/src/lib.rs [hunk 4]
```diff
@@ -333,9 +346,30 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
...
-				let origin = T::EVMBridge::get_origin().ok_or(Error::<T>::RealOriginNotFound)?;
-				let origin_address = T::AddressMapping::get_or_create_evm_address(&origin);
+				let erc20_holding_account = T::AddressMapping::get_account_id(&T::Erc20HoldingAccount::get());
+				let origin_address = if erc20_holding_account == to.clone() {
+					// withdraw **from** sender to erc20 holding account. in xcm case which receive erc20 from sibling
+					// parachain, sender is sibling parachain sovereign account. As the origin here is used to charge
+					// storage fee, we must make sure sibling parachain sovereign account has enough token to charge
+					// storage fee.
+					T::AddressMapping::get_or_create_evm_address(from)
+				} else if erc20_holding_account == from.clone() {
```
Comment (zjb0807) [inline modules/currencies/src/lib.rs:356]: Why not do this in `deposit` and `withdraw`? The code will be more clearer.

File: modules/evm/src/lib.rs (hunk 3 above)
Comment (zqhxuyuan) [reply modules/evm/src/lib.rs:1780]: remove it now. previous I'm using it in testcase to manual clear origin.

File: modules/currencies/src/lib.rs [hunk 5]
```diff
@@ -364,12 +364,35 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
...
+				// parachain, we choose receiver to charge storage fee. we must make sure receiver has enough native
+				// token to charge storage fee.
+				let from = T::AddressMapping::get_account_id(&T::Erc20HoldingAccount::get());
+				let sender = T::AddressMapping::get_evm_address(&from).ok_or(Error::<T>::EvmAccountNotFound)?;
+				ensure!(
+					!Self::free_balance(currency_id, &from).is_zero(),
+					Error::<T>::RealOriginNotFound
+				);
+				let origin = T::AddressMapping::get_or_create_evm_address(who);
+				let receiver = T::AddressMapping::get_or_create_evm_address(who);
```
Comment (zjb0807) [inline modules/currencies/src/lib.rs:378]:
```suggestion
				let receiver = T::AddressMapping::get_or_create_evm_address(who);
				let origin = receiver;

File: modules/currencies/src/lib.rs [hunk 6]
```diff
@@ -364,12 +364,34 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
...
-			),
+			CurrencyId::Erc20(contract) => {
+				// deposit from erc20 holding account to receiver(who). in xcm case which receive erc20 from sibling
+				// parachain, we choose receiver to charge storage fee. we must make sure receiver has enough native
+				// token to charge storage fee.
+				let sender = T::Erc20HoldingAccount::get();
+				let from = T::AddressMapping::get_account_id(&sender);
+				ensure!(
+					!Self::free_balance(currency_id, &from).is_zero(),
+					Error::<T>::RealOriginNotFound
```
Comment (xlc) [inline modules/currencies/src/lib.rs:375]: why `RealOriginNotFound` error when free balance is zero?

File: modules/currencies/src/lib.rs [hunk 7]
```diff
@@ -364,12 +364,34 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
...
+				T::EVMBridge::transfer(
+					InvokeContext {
+						contract,
+						sender,
+						origin: receiver,
+					},
+					receiver,
+					amount,
+				)?;
+				Self::deposit_event(Event::Transferred {
```
[the remark this answers is not in the document]
Comment (zqhxuyuan) [reply modules/currencies/src/lib.rs:387]: should `Deposited` only effect on ERC20 case, or all currency_id, other currency_id already has event in Tokens(multi_currency) or Balances(native_currency) module.

[the remark this answers is not in the document]
Comment (shaunxw) [reply modules/currencies/src/lib.rs:387]: Currently all other non-erc20 currencies will emit `Deposited` in this `deposit` fn. Making it `Transferred` only for erc20 is not consistent either.

[the remark this answers is not in the document]
Comment (zqhxuyuan) [reply modules/currencies/src/lib.rs:387]: Do we need `Withdrawn(erc20_holding_account)` event in `deposit()` fn?

**Review by shaunxw — APPROVED** [review]

**Pull request merged**